### PR TITLE
[RLlib] Meta-learning, differentiable learner.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Implementing meta-learning using our default `TorchLearner` can be cumbersome, particularly when dealing with second-order or higher-order gradients. Meta-learning requires functional parameter updates, allowing parameters to be reused in subsequent meta-updates.

This PR introduces a refined design - building upon the existing framework - by introducing a `TorchMetaLearner`. This new component incorporates one or more `DifferentiableLearner` instances, where each `DifferentiableLearner` performs functional updates, consistently returning cloned and updated parameters. The `TorchMetaLearner` then utilizes these parameters to facilitate meta-updates, enabling higher-order gradient computations.

Additionally, this PR includes a concise example demonstrating how to implement Model-Agnostic Meta-Learning (MAML) in just a few lines of code while successfully replicating results from the original paper in a supervised learning setting.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm_config.py [hunk 1]
```diff
@@ -5741,6 +5744,119 @@ def num_cpus_for_local_worker(self, value):
...
+                config.
+        """
+        # Initialize the `AlgorithmConfig` first.
+        super().__init__(algo_class=algo_class)
+
+        # Initialize the list of differentiable learner configs to an empty list, which
+        # defines the default, i.e. the `MetaLearner` will have no nested updates.
+        self.differentiable_learner_configs: List[DifferentiableLearnerConfig] = []
+
+    def learners(
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:5795]: Note that `learners` is also a superclass method, so we should accept **kwargs here and pass these to the super method first:
```
def learners(self, *, a, **kwargs):
    super().learners(**kwargs)
    
    self.a = a
    return self
```

Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:5795]: nit: add override decorator

File: rllib/algorithms/algorithm_config.py [hunk 2]
```diff
@@ -5741,6 +5744,119 @@ def num_cpus_for_local_worker(self, value):
...
+            differentiable_learner_configs: A list of `DifferentiableLearnerConfig` instances
+                defining the `DifferentiableLearner` classes used for the nested updates in
+                `Algorithm`'s learner.
+        """
+        if differentiable_learner_configs is not NotProvided:
+            self.differentiable_learner_configs = differentiable_learner_configs
+
+        return self
+
+    def validate(self):
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:5812]: nit: add override decorator

File: rllib/algorithms/algorithm_config.py [hunk 3]
```diff
@@ -5741,6 +5744,119 @@ def num_cpus_for_local_worker(self, value):
...
+            isinstance(learner_cfg, DifferentiableLearnerConfig)
+            for learner_cfg in self.differentiable_learner_configs
+        ):
+            self._value_error(
+                "`differentiable_learner_configs` must be a list of "
+                "`DifferentiableLearnerConfig` instances, but at least "
+                "one instance is not a `DifferentiableLearnerConfig`."
+            )
+
+    def get_default_learner_class(self):
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:5846]: nit: add override decorator

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+# Import torch.
+torch, _ = try_import_torch()
+
+# Define arguments.
+parser = add_rllib_example_script_args(
+    default_iters=70_000,
+)
+
+parser.add_argument(
+    "--train-batch-size",
```
Comment (sven1977) [inline rllib/examples/learners/maml_lr_supervised_learning.py:114]: We should name these something with `meta`. We may add the command line args `--train-batch-size` and `--lr` in the future to set the regular (non-meta) config values, so it's better to be explicit here.

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+    type=int,
+    default=5,
+    help="The number of samples per train and test update (meta-learning).",
+)
+parser.add_argument(
+    "--lr",
```
Comment (sven1977) [inline rllib/examples/learners/maml_lr_supervised_learning.py:120]: same: add `meta` to name of arg.

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+parser.add_argument(
+    "--fine-tune-batch-size",
+    type=int,
+    default=10,
+    help="The number of samples for the fine-tuning updates.",
+)
+parser.add_argument(
+    "--noise-std",
+    type=float,
+    default=0.0,
```
Comment (sven1977) [inline rllib/examples/learners/maml_lr_supervised_learning.py:134]: Should the default be 0.0 here? Or 1.0?

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    "--pause-plot-secs",
+    type=int,
+    default=1000,
+    help=(
+        "The number of seconds to keep the plot open. Note the plot can always be "
+        "closed by the user when open."
+    ),
+)
+
+# Parse the arguments.
```
Comment (sven1977) [inline rllib/examples/learners/maml_lr_supervised_learning.py:173]: add a `if __name__ == "__main__":`

block here and indent everything below into this block (except the helper functions below, which should be moved above the `parser` definition.

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+args = parser.parse_args()
+
+# If a random seed is provided set it for torch and numpy.
+if args.seed:
+    torch.random.manual_seed(args.seed)
+    np.random.seed(args.seed)
+
+# Implement generation of data from sinusoid curves.
+def generate_sinusoid_task(batch_size, noise_std=0.1, return_params=False):
```
Comment (sven1977) [inline rllib/examples/learners/maml_lr_supervised_learning.py:182]: move this function above `parser` definition to head of file.

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+        )
+    # Otherwise, return only the sampled data.
+    else:
+        return (
+            torch.tensor(x, dtype=torch.float32),
+            torch.tensor(y, dtype=torch.float32),
+        )
+
+
+def sample_task(batch_size=10, noise_std=0.1, training_data=False, return_params=False):
```
Comment (sven1977) [inline rllib/examples/learners/maml_lr_supervised_learning.py:226]: move this function above `parser` definition to head of file.

File: rllib/core/learner/torch/torch_meta_learner.py [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+    @OverrideToImplementCustomLogic
+    @override(Learner)
+    def update(
+        self,
+        batch: Optional[MultiAgentBatch] = None,
+        batches: Optional[List[MultiAgentBatch]] = None,
+        batch_refs: Optional[List[ray.ObjectRef]] = None,
+        episodes: Optional[List[EpisodeType]] = None,
+        episodes_refs: Optional[List[ray.ObjectRef]] = None,
+        data_iterators: Optional[List[ray.data.DataIterator]] = None,
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_meta_learner.py:106]: Happy to already NOT support these helper args anymore and go all-in on `TrainingData`. At least for this new class here.

File: rllib/core/learner/torch/torch_meta_learner.py [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+        *,
+        # TODO (simon): Maybe defined here other_training_data to provide
+        # TrainingData for differentiable Learners. List to nest them.
+        # other_training_data: List[Union[List[Any], TrainingData]]
+        # TODO (sven): Make this a more formal structure with its own type.
+        timesteps: Optional[Dict[str, Any]] = None,
+        num_total_minibatches: int = 0,
+        num_epochs: int = 1,
+        minibatch_size: Optional[int] = None,
+        shuffle_batch_per_epoch: bool = False,
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_meta_learner.py:117]: I think we can get rid of these, too, and replace them with `self.config.num_epochs` ...

**Review by sven1977 — APPROVED** [review]
Awesome PR @simonsays1980 !! Thanks for all the work.

A few high-level items:
a) Can we add the example so it runs in the CI?
b) We need to add the example to the docs. I would actually move all the MAML example files (main and classes) into the `examples/algorithms` sub-folder.
c) Maybe for a follow-up PR. How hard would it be to make MAML an actual RLlib Algorithm?

File: rllib/algorithms/algorithm_config.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm_config.py:5795]: Yup, I ran over this in my meta learning algorithm implementation.

File: rllib/examples/learners/maml_lr_supervised_learning.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/examples/learners/maml_lr_supervised_learning.py:134]: Should be `0.0`. We do not want to make it harder than it is already (note: we are basically learning to adapt to infinitely many tasks in few-shot).

File: rllib/core/learner/torch/torch_meta_learner.py (hunk 10 above)
Comment (simonsays1980) [reply rllib/core/learner/torch/torch_meta_learner.py:106]: Happy about this, too!

File: rllib/core/learner/torch/torch_meta_learner.py (hunk 11 above)
Comment (simonsays1980) [reply rllib/core/learner/torch/torch_meta_learner.py:117]: Yeah, let's do this.

File: rllib/examples/learners/maml_lr_supervised_learning.py [hunk 12]
```diff
@@ -0,0 +1,423 @@
+"""Example of how to use `TorchMetaLearner` and `DifferentiableLearner` for MAML.
+
+In this MAML example the model is trained through meta learning to be able to adapt
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/examples/learners/maml_lr_supervised_learning.py:3]: The algorithm is a SL algorithm here - the RL example is way more cumbersome and will come after the implementation of the meta algo we are heading for. Will add more precise descriptions about the problem. 

Do you think, we really should describe how meta learning works? Or better how meta learning works with RLlib?

**Comment by simonsays1980** [comment]
> Awesome PR @simonsays1980 !! Thanks for all the work.
> 
> A few high-level items: a) Can we add the example so it runs in the CI? b) We need to add the example to the docs. I would actually move all the MAML example files (main and classes) into the `examples/algorithms` sub-folder. c) Maybe for a follow-up PR. How hard would it be to make MAML an actual RLlib Algorithm?

Thanks @sven1977! I answer to your quote below:

a. Yes we can do this. I will simply block plotting when `--as-test` is used.
b. I will add this example to the docs as well. Will wait until the meta-learning algorithm I am working on is ready. There will come a couple of additional features needed for meta learning. Your suggestions from this PR will come in another one that I will provide this week.
c. Its harder than SL meta learning, but I think it will work fine. Its on my task list for after the offline meta learning algo is done. Great suggestion! I think this would be awesome.
